src/addrdb.cpp

```diff
@@ -161,13 +161,13 @@ bool CAddrDB::Read(CAddrMan& addr, CDataStream& ssPeers)
 void DumpAnchors(const fs::path& anchors_db_path, const std::vector<CAddress>& anchors)
 {
     LOG_TIME_SECONDS(strprintf("Flush %d outbound block-relay-only peer addresses to anchors.dat", anchors.size()));
-    SerializeFileDB("anchors", anchors_db_path, anchors);
+    SerializeFileDB("anchors", anchors_db_path, anchors, ADDRV2_FORMAT);
 }
 
 std::vector<CAddress> ReadAnchors(const fs::path& anchors_db_path)
 {
     std::vector<CAddress> anchors;
-    if (DeserializeFileDB(anchors_db_path, anchors)) {
+    if (DeserializeFileDB(anchors_db_path, anchors, ADDRV2_FORMAT)) {
```

Comment:
I think this will not deserialize properly `anchors.dat` from before this PR.
* `CLIENT_VERSION` is something like `219900`
* `ADDRV2_FORMAT` is `0x20000000`

In `anchors.dat` before this PR we have `CAddress` entries which contain `version=219900` and addresses serialized in addrv1 format.

With this PR we will set the stream version to `ADDRV2_FORMAT` during [deserialize](https://github.com/bitcoin/bitcoin/blob/50091592dd875a1c94030dbed74112b003732d68/src/protocol.h#L369-L395):

```cpp
    // Notice added comments below
    SERIALIZE_METHODS(CAddress, obj)
    {
        SER_READ(obj, obj.nTime = TIME_INIT);
        int nVersion = s.GetVersion(); // nVersion will be assigned ADDRV2_FORMAT (0x20000000)
        if (s.GetType() & SER_DISK) {
            READWRITE(nVersion); // nVersion will be assigned CLIENT_VERSION (219900)
        }
...
        if (nVersion & ADDRV2_FORMAT) { // this will be false
            uint64_t services_tmp;
            SER_WRITE(obj, services_tmp = obj.nServices);
            READWRITE(Using<CompactSizeFormatter<false>>(services_tmp));
            SER_READ(obj, obj.nServices = static_cast<ServiceFlags>(services_tmp));
        } else {
            // it will read services as 8 bytes (correct, as that is what is actually stored on disk)
            READWRITE(Using<CustomUintFormatter<8>>(obj.nServices));
        }
        // oops, this will deserialize CService, which will deserialize CNetAddr which will check
        // s.GetVersion() which is ADDRV2_FORMAT (0x20000000) and will deserialize the
        // address as if it is stored in addrv2 format on disk, but it is actually stored in addrv1,
        // so this will deserialize garbage
        READWRITEAS(CService, obj);
    }
```

Reply:
I think in this case the file checksum will mismatch and we won't load any anchors at all, which is ok (anchors.dat is not in any release and losing the file isn't a problem).

Reply:
My point is that it will not be able to read "old" `anchors.dat` (I am not sure of the exact way it will fail, maybe it will be indeed checksum failure).

I mention this because the comments above give the impression that "old" `anchors.dat` would be readable.

Reply:
> This should be compatible with (v1) addresses stored by 0.21.0rc{1,2} as well...

Such compatibility shouldn't be a point for the reason mentioned by @jnewbery.